avocado/plugins/runner.py

```diff
@@ -324,37 +324,34 @@ def _template_to_factory(test_parameters, template, variant):
                           "variant_id": varianter.generate_variant_id(var),
                           "paths": paths}
 
-    def _iter_suite(self, job, test_suite, variants, execution_order):
+    def _iter_suite(self, job, variants, execution_order):
         """
         Iterates through test_suite and variants in defined order
 
         :param job: an instance of :class:`avocado.core.job.Job`
-        :param test_suite: a list of tests to run
         :param variants: a varianter object to produce test params
         :param execution_order: way of iterating through tests/variants
         :return: generator yielding tuple(test_factory, variant)
         """
         if execution_order == "variants-per-test":
             return (self._template_to_factory(job.test_parameters,
                                               template, variant)
-                    for template in test_suite
+                    for template in job.test_suite
                     for variant in variants.itertests())
         elif execution_order == "tests-per-variant":
             return (self._template_to_factory(job.test_parameters,
                                               template, variant)
                     for variant in variants.itertests()
-                    for template in test_suite)
+                    for template in job.test_suite)
         else:
             raise NotImplementedError("Suite_order %s is not supported"
                                       % execution_order)
 
-    def run_suite(self, job, result, test_suite, variants):
+    def run_suite(self, job, variants):
         """
         Run one or more tests and report with test result.
 
         :param job: an instance of :class:`avocado.core.job.Job`.
-        :param result: an instance of :class:`avocado.core.result.Result`
-        :param test_suite: a list of tests to run.
```

Comment:
How do you plan to have such an implementation run one specific test suite, out of many that may exist in a job in the future (as per #3958) ?

Reply:
The test suite (or suites) will be part of a Job:

```python3
suite = Suite(config)
job = Job([suite])
job.run() # that will call runner.run_job(self)
```

This method could run over a loop with all test suites and call a specific test suite.

Makes sense?

Reply:
Well, then the interface changes completely, and it's no longer about running *a* test suite, but running everything a job has.

There are implications with that, that we must be aware.  For instance, it may block us from running *suites* independently (without the required knowledge of each implementation).  IMO, this goes in the opposite direction of some use cases we thought about.  It prevents, for instance, a Job writer from running one test suite, and depending on its results, proceeding or not with another suite (this is a very common use case).

Reply:
@clebergnu I see your point and I'm not planning to remove the ability to run one specific test suite. We can have both `run_job()` and `run_suite()`. So we will still have the ability to run one specific test suite.

This change is only to make sense semantically since that `run_suite(job)` for now sounds strange to me. I know that we are running a suite, but we are passing a job.

I have intentions to bring back the `run_suite()` method but with a proper "test suite" object as an argument.

Reply:
BTW, this is very old resource (RFC) about the Job API:

https://www.redhat.com/archives/avocado-devel/2016-April/msg00014.html

But it's probably an interesting read even now.  About some of the points I mentioned before, this is one that maps very close:

---

Ordered and conditional test execution

A user wants to create a custom job that only runs a benchmark test on
a VM if the VM installation test succeeds.

Possible use case fulfillment

Pseudo code:

```python
  #!/usr/bin/env python
  from avocado import Job
  from avocado.resolver import resolve

  job = Job()

  vm_install = resolve('io-github-autotest-qemu.unattended_install.cdrom.http_ks.default_install.aio_native')
  vm_disk_benchmark = resolve('io-github-autotest-qemu.autotest.bonnie')

  if job.run_test(vm_install).result == 'PASS':
      job.run_test(vm_disk_benchmark)
```

Just think of `run_test_suite()` instead of `run_test()`.

Reply:
> @clebergnu I see your point and I'm not planning to remove the ability to run one specific test suite. We can have both `run_job()` and `run_suite()`. So we will still have the ability to run one specific test suite.
> 

The way I see this, a job is an active entity, that is pretty much in charge of its state, so, I find it wrong to have a `run_job()` method implemented by a plugin.  I mean, what's the scope of such an interface and its responsibilities?  How much of the "job" work should it do?  I maybe just the naming, but there's a lack of precision on the goal of the interface here IMO.

Now, if you want to call it `run_all_of_the_job_test_suites()`, then fine, the scope is pretty much well described.  But, at the same time, what's the point of having that in an external plugin, and not in the base job class?

> This change is only to make sense semantically since that `run_suite(job)` for now sounds strange to me. I know that we are running a suite, but we are passing a job.
> 

That's only because we have not attempted to limit what the plugins can do with the job.  IMO it's much more honest to change the order of the parameters, say:

```python
  run_suite(suite, job)
```

Then to say that it's really running a job (there's more to a job than that).

> I have intentions to bring back the `run_suite()` method but with a proper "test suite" object as an argument.

Which would probably mean giving the reinstated `run_suite()` both a `suite` *and* a `job`, which gets back to the change of parameter order I mentioned before.

Reply:
> > @clebergnu I see your point and I'm not planning to remove the ability to run one specific test suite. We can have both `run_job()` and `run_suite()`. So we will still have the ability to run one specific test suite.
> 
> The way I see this, a job is an active entity, that is pretty much in charge of its state, so, I find it wrong to have a `run_job()` method implemented by a plugin. I mean, what's the scope of such an interface and its responsibilities? How much of the "job" work should it do? I maybe just the naming, but there's a lack of precision on the goal of the interface here IMO.

Ok, Agreed.

> Now, if you want to call it `run_all_of_the_job_test_suites()`, then fine, the scope is pretty much well described. But, at the same time, what's the point of having that in an external plugin, and not in the base job class?
> 
> > This change is only to make sense semantically since that `run_suite(job)` for now sounds strange to me. I know that we are running a suite, but we are passing a job.
> 
> That's only because we have not attempted to limit what the plugins can do with the job. IMO it's much more honest to change the order of the parameters, say:
> 
> ```python
>   run_suite(suite, job)
> ```

This makes no sense to me because job will have the "suite" or "suites" parameter. We are passing the suite twice.

Reply:
> > That's only because we have not attempted to limit what the plugins can do with the job. IMO it's much more honest to change the order of the parameters, say:
> > ```python
> >   run_suite(suite, job)
> > ```
> 
> This makes no sense to me because job will have the "suite" or "suites" parameter. We are passing the suite twice.

But the question remains: how will you pick *one* suite from the job's suites? That's the missing bit.

Reply:
@clebergnu well, lets say that `job.suites` is a list of `TestSuite()` objects. (To be Implemented). And each `TestSuite` object has an `id` or a `name`. I think that would be easy to implement something like `job.get_suite_by_id(self, id)` or `job.get_suite_by_name(self, name)`, right?

At first, I think on a list because I don't see any peformance issue for now, but could be a dict when peformance becomes a problem.

Not sure if this is the missing bit.